lib/concurrent/lazy_reference.rb

```diff
@@ -0,0 +1,62 @@
+module Concurrent
+
+  # Lazy evaluation of a block yielding an immutable result. Useful for
+  # expensive operations that may never be needed. `LazyReference` is a simpler,
+  # blocking version of `Delay` and has an API similar to `AtomicReference`.
+  # The first time `#value` is called the caller will block until the
+  # block given at construction is executed. Once the result has been
+  # computed the value will be immutably set. Any exceptions thrown during
+  # computation will be suppressed.
+  #
+  # Because of its simplicity `LazyReference` is much faster than `Delay`:
+  #
+  #            user     system      total        real
+  #     Benchmarking Delay...
+  #        0.730000   0.000000   0.730000 (  0.738434)
+  #     Benchmarking LazyReference...
+  #        0.040000   0.000000   0.040000 (  0.042322)
+  #
+  # @see Concurrent::Delay
+  class LazyReference
+
+    # Creates anew unfulfilled object.
+    #
+    # @yield the delayed operation to perform
+    # @param [Object] default (nil) the default value for the object when
+    #   the block raises an exception
+    #
+    # @raise [ArgumentError] if no block is given
+    def initialize(default = nil, &block)
+      raise ArgumentError.new('no block given') unless block_given?
+      @default = default
+      @task = block
+      @mutex = Mutex.new
+      @value = nil
+      @fulfilled = false
+    end
+
+    # The calculated value of the object or the default value if one
+    # was given at construction. This first time this method is called
+    # it will block indefinitely while the block is processed.
+    # Subsequent calls will not block.
+    #
+    # @return [Object] the calculated value
+    def value
+      # double-checked locking is safe because we only update once
+      return @value if @fulfilled
```

Comment:
I'm all for making Delay faster, but that's not the primary reason I created LazyReference. I created LazyReference because the basic use case for Delay is no longer easily or intuitively supported. The initial use case for Delay was to lazy load a resource in such a way that the first call to `value` would block the current thread *indefinitely* and execute the block *on that thread*:

```ruby
# original implementation
d = Concurrent::Delay.new{ sleep 10; 'foo' } # no constructor parameters
d.value # blocks the current thread and executes the block on the current thread
```

The current behavior of Delay is different:

```ruby
# current implementation
d = Concurrent::Delay.new{ sleep 10; 'foo' } # no constructor parameters
d.value # blocks the current thread and executes the block on a *global thread pool*
```

Notice that the current version of Delay blocks *2* threads: The current thread and the thread on the global thread pool that executes the block. This is very different behavior and also the source of the bug with the root actor (which we originally thought was a problem with autotest).

To get the same behavior as the *original* Delay implementation we must pass constructor arguments:

```ruby
# current implementation
d = Concurrent::Delay.new(executor: Concurrent::ImmediateExecutor.new){ sleep 10; 'foo' } # passing an executor into the constructor
d.value # blocks the current thread and executes the block on the current thread
```

This syntax is not every obvious nor very intuitive and is easy to get wrong. If we, the authors of this gem, got it wrong with the root actor, it's reasonable to assume that our users will get it wrong in critical situations, too.

What's most interesting about this base case is that every use of Delay within this gem follows the ImmediateExecutor pattern. Every one. Moreover, I was able to replace every use of Delay within this gem with LazyReference without losing functionality (with the exception of the root actor, which remains a Delay). That strongly indicates that this base use case is the most important use case and, therefore, the use case that should be supported by the simplest use of the abstraction. LazyReference does this, Delay does not.